### ISSUE TITLE
Web API c#.net Response 500 null reference json serialization

### ISSUE THREAD

**Issue by SDNADEV** [JamesNK/Newtonsoft.Json#2545]
Hi Team,

We are using web APIs in our project, The time we access the first web API from our project we get a 500 error as a response. once we restart the app pool on the machine, the error goes away and everything works fine.

It occurs very rarely. So, it's very hard to reproduce. error is invoked on its own

As per our current investigation, we noted once in our local machine while debugging found the attached exception.

We get a null reference exception while serializing objects to JSON. check attachment for info

The issue occurs very rarely on both server(Windows Server 2019) and local window machine
.net framework version being used with this MVC c#.net web API is 4.5.1

Following things, we have tried: -

Upgraded the json version to Newtonsoft.json.13.0.1

Update the response to a MVC model from as earlier we were sending entity

Globally we have setup the below line in webapiconfig file
config.Formatters.JsonFormatter.SerializerSettings.ReferenceLoopHandling = Newtonsoft.Json.ReferenceLoopHandling.Ignore;
![Error500 ](https://user-images.githubusercontent.com/85789756/121774144-77cb1000-cb9e-11eb-922c-e41d1b1d1128.PNG)

**Comment by elgonzo** [comment]
Does your project by any chance configure and/or use JsonConverters somewhere?

From what i can tell, a NullReferenceException directly originating in the _JsonSerializer.GetMatchingConverter_ method https://github.com/JamesNK/Newtonsoft.Json/blob/e44ab3348fabcbe9ed6b5053b706a9e10b945ba1/Src/Newtonsoft.Json/JsonSerializer.cs#L1199-L1219
should only happen if a `null` value has been added to the converter collection used by the serializer...

**Comment by shivanshub** [comment]
I downloaded the code and trying to connect it. This error is random and so hard to reproduce; go away with app pool restart. For now, It seems like some thread safety issue, maybe in the library code or in the project code.

**Comment by SDNADEV** [comment]
> Does your project by any chance configure and/or use JsonConverters somewhere?
> 
> From what i can tell, a NullReferenceException directly originating in the _JsonSerializer.GetMatchingConverter_ method
> 
> https://github.com/JamesNK/Newtonsoft.Json/blob/e44ab3348fabcbe9ed6b5053b706a9e10b945ba1/Src/Newtonsoft.Json/JsonSerializer.cs#L1199-L1219
> 
> 
> should only happen if a `null` value has been added to the converter collection used by the serializer...

Yes, we are using customJsonConverter in our code and adding them in Global.asax.cs file

**Comment by hemiaoio** [comment]
I also got the same error and I can't reproduce it, any progress now?
